flux-desugar/src/zip_resolver.rs

```diff
@@ -90,26 +98,41 @@ fn zip_ty(ty: Ty, rust_ty: &rustc_ty::Ty) -> Ty<Res> {
     Ty { kind, span: ty.span }
 }
 
-fn zip_path(path: Path, rust_ty: &rustc_ty::Ty) -> Path<Res> {
+fn zip_path(tcx: TyCtxt, path: Path, rust_ty: &rustc_ty::Ty) -> Path<Res> {
     let (res, rust_args) = rustc_ty_ident_args(rust_ty);
 
     let path_args_len = path.args.len();
-    if path_args_len != rust_args.len() {
-        panic!("type-argument mismatch, expected: {:#?},  found: {:#?}", rust_args, path.args);
-    }
-    let args: Vec<Ty<Res>> = iter::zip(path.args, rust_args)
-        .map(|(arg, rust_arg)| zip_generic_arg(arg, rust_arg))
+    // Assume that the rust_args are of the form [path_args + default_args]
+    // i.e. default args all come _after_ the supplied path_args.
+    let rust_args_len = rust_args.len();
+    let default_args_len = rust_args_len - path_args_len;
+    assert!(default_args_len <= rust_args_len);
+
+    // zip the supplied args
+    let mut args: Vec<Ty<Res>> = iter::zip(path.args, rust_args)
+        .map(|(arg, rust_arg)| zip_generic_arg(tcx, arg, rust_arg))
         .collect();
 
+    // copy over the 'default' args with trivial refinements
+    if 0 < default_args_len {
+        match res {
+            Res::Adt(def_id) => {
+                let generics = &tcx.generics_of(def_id).params[path_args_len..];
```

Comment:
I feel it'd be better if we add the missing default parameters in the lowering (`core::Ty -> ty::Ty`) and here we just resolve the parameters we do have. The reason I think it'd be better is that in the lowering we can use https://github.com/liquid-rust/flux/blob/main/flux-middle/src/global_env.rs#L207 to get the trivially refined type. But I'm not sure if that will cause any other problems, so I'm ok with revisiting it later.

In general, I'm not happy with generating default types in `surface` or `core`.

Reply:
There are 3 options

1. Add the params here (i.e. `zip_resolver`)
2. Ignore the params and just deal with the mismatched args as the appear
3. Add the params later (e.g. in `lowering`)

(2) seems dicey and a bit ad-hoc (we might miss some cases and cause odd silent failures.)

(3) is challenging at that point (`lowering`) we don't have the original `rustc` signature so we 
don't know what extra params to add.

Hence (1) seemed the least worst for now...

Reply:
I think 3 can be done, but getting the default type is more circuitous. I [found this today](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/generics/struct.GenericParamDef.html#method.default_value) (I randomly read rustc source in my free time) which I think will serve the purpose. But I'm not terribly convinced this will end up being cleaner or that the function I found will get us what we need.

Reply:
OK let me add a link to this discussion in the code, and we can revisit later. [ IMO doing it in 3 is going to duplicate a bunch of the resolving work... ]